### PR TITLE
Optimize CI file structure by moving redundant steps into a common action

### PR DESCRIPTION
**Pull request by tanertopal**
### Description

The CI workflows have a lot of redundant definitions, which make updating error prone and hard to debug in case something goes wrong. This PR adds an action which merges the repetitive preparation steps into one and will make adding caching actions down the line easier.

### Checklist

- [x] Implement proposed change
- [x] Make CI checks pass
- [x] Ping maintainers on [Slack](https://flower.dev/join-slack/) (channel `#contributions`)

### CODE REVIEW

File: .github/actions/bootstrap/action.yml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+name: "bootstrap"
+description: "Bootstrap Python environment including Python dependencies and configure optimizations."
+inputs:
+  python-version:
+    description: "Version range or exact version of Python or PyPy to use, using SemVer's version range syntax."
```
Comment (danieljanes) [inline .github/actions/bootstrap/action.yml:5]: Why PyPy?

File: .github/actions/bootstrap/action.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
+name: "bootstrap"
+description: "Bootstrap Python environment including Python dependencies and configure optimizations."
```
Comment (danieljanes) [inline .github/actions/bootstrap/action.yml:2]:
```suggestion
description: "Bootstrap Python environment (install and configure Python version and build tools)"
```

File: .github/workflows/e2e.yml [hunk 3]
```diff
@@ -17,16 +17,8 @@ jobs:
...
-        with:
-          python-version: 3.7.15
-      - name: Install
-        run: |
-          python -m pip install -U pip==22.3.1
-          python -m pip install -U setuptools==65.6.3
-          python -m pip install poetry==1.3.2
-          poetry config virtualenvs.create false
+      - name: Bootstrap
+        uses: ./.github/actions/bootstrap
```
Comment (danieljanes) [inline .github/workflows/e2e.yml:21]: This implicitly changes Python from 3.7 to 3.8

File: .github/workflows/e2e.yml [hunk 4]
```diff
@@ -258,16 +202,8 @@ jobs:
     timeout-minutes: 10
     steps:
       - uses: actions/checkout@v3
-      - name: Set up Python
-        uses: actions/setup-python@v4
-        with:
-          python-version: 3.8.17
```
Comment (danieljanes) [inline .github/workflows/e2e.yml:264]: Why is the default for the new bootstrap action 3.8.16 if .17 is available?

**Review by danieljanes — CHANGES_REQUESTED** [review]

File: .github/actions/bootstrap/action.yml (hunk 1 above)
Comment (tanertopal) [reply .github/actions/bootstrap/action.yml:5]: @danieljanes this is just passed through to the office action and the description is a copy.

File: .github/workflows/e2e.yml (hunk 3 above)
Comment (tanertopal) [reply .github/workflows/e2e.yml:21]: I'll update it to .17. 

As a sidenote in the improve_ci_speed PR there will be some more changes and explanations on that.

File: .github/workflows/e2e.yml (hunk 4 above)
Comment (tanertopal) [reply .github/workflows/e2e.yml:264]: I'll update it to .17. 

As a sidenote in the improve_ci_speed PR there will be some more changes and explanations on that.

File: .github/workflows/e2e.yml (hunk 3 above)
Comment (danieljanes) [reply .github/workflows/e2e.yml:21]: It's often good to separate refactoring changes ("moving stuff around") and actual changes (changing how something behaves) into separate PR. A refactoring should not change any behaviour, which also makes it easier to verify that the refactoring is correct. The actual change (in this case: changing the version of those CI jobs) is trivial later on and should also be done in isolation. If we find an issue later on, this makes it easier to identify whether the issue has been introduced by the refactoring or by the version change.

TL;DR: let's stick with 3.7 in this PR

Comment (tanertopal) [reply .github/workflows/e2e.yml:21]: Not sure I agree in this case as we are modifying the CI jobs but I'll revert the changes so each uses its own previous Python version and open PR's afterwards for each job.

Comment (danieljanes) [reply .github/workflows/e2e.yml:21]: If we don't change the Python version, the DRY part of this PR is a typical refactoring. It just "moves code around", but the CI behaviour should not change. In fact, if it changes, it's a bug. The only small exception is the Pandoc installation because there the installation order changes.

Comment (tanertopal) [reply .github/workflows/e2e.yml:21]: @danieljanes, all functional changes are reverted now.

**Review by danieljanes — APPROVED** [review]
